Loading all Markdown in memory

d0ugal opened the pull request:
This change updates the Markdown loading to load all the documents at startup. This gives us much more flexibility when we want to reference other Markdown files or use the metadata from a Markdown file earlier in the build.

Remaining tasks:

- [ ] Fix a few test  failures on older Python versions
- [ ] Fix the code in a few places that is wonky. Commented on inline.
- [ ] Add to release notes

Original text below the line.

---

This needs a bunch of works and it breaks loads of the tests, but it works quite well. It gets the page title from the Markdown if it isn't provided in the mkdocs.yml.

Feedback very welcome, it is a bit rough at the moment tho'

See #316

waylan commented:
Looks like a good start.

One thing that concerns me is that the Markdown is rendered to HTML too soon. If/when a plugin API is added, I anticipate a hook being provided prior to each page being built which means the HTML should not be rendered until build time. I realize that it is rendered early because that is how you are getting the meta-data for each page. And yes, that info should already be available before the pre-build hook.

Of course, those two things are in direct opposition to each other. However, I recommend that the Python-Markdown's Meta-Data Extension not be used. Instead extract the meta-data early (perhaps using [docdata][1]), and save the meta-data and the markdown source (with the meta-date removed) as two separate attributes in the Page instance. The way, the Markdown source can be used later to render the HTML just before building the page, but after the pre-build hook.

If you don't want to add another dependency, I give permission to include the relevant parts of docdata directly in MkdDocs. Or you are welcome to use some other implementation of your choice or creation.

An interesting side-affect will be that the code which calls Markdown will still be in `build.py` so you won't need to move the Markdown wrapper (leave it in `build.py`) and you won't need to update those tests (which are many of the currently failing tests).

[1]: https://github.com/waylan/docdata

facelessuser commented:
> However, I recommend that the Python-Markdown's Meta-Data Extension not be used.

+1 to this.  Personally, I have never found the meta extension to be all that useful as I want the meta data **before** I parse.  Usually, by time I've parsed is too late for me to use the meta for what I want.

d0ugal commented:
Great, I'm glad you both said this because I was struggling with it too. It did indeed feel a bit strange to render the Markdown so early.

> If you don't want to add another dependency, I give permission to include the relevant parts of docdata directly in MkdDocs. Or you are welcome to use some other implementation of your choice or creation.

@waylan Great - thanks! I am a bit hesitant to include another dependency, just because of the effort recently we had to do to remove two (six and ghp-import). So I might take this approach, initially I'll try a prototype with the dependency and see how that goes.

d0ugal commented:
Updated. I think this generally does make much more sense. Unfortunately, most of the tests still break as they are broken due to how I changed the internal pages structure. This PR changes it so that we convert from string paths to page objects as soon as possible. The previous was was quite delayed and the tests took advantage of this to easily define a set of pages as a list of strings.

d0ugal commented:
... and because we load the pages when we do now, it verifies they all exist. A bunch of tests have fake paths to markdown files because they never tested the build process. /me *sighs*

waylan commented on mkdocs/build.py:
```diff
@@ -164,30 +166,16 @@ def build_template(template_name, env, config, site_navigation=None):
     context = get_global_context(site_navigation, config)
     context.update(get_page_context(
-        page, html_content, table_of_contents, meta, config
+        page, html_content, toc, page.meta, config
```
Within the template I would expect the html_content, toc, and meta to all be attributes of the page object. I realize that may involve so template refactoring and this is prototyping right now, but I think that is we're this needs to go.

d0ugal replied:
Yeah, I agree. In my previous version I had added it to the template, but left it available via the old names too to avoid updating all the templates at the same time.

d0ugal commented on mkdocs/nav.py:
```diff
@@ -26,6 +29,37 @@ def filename_to_title(filename):
+                                            use_directory_urls, docs_dir,
+                                            title=next_cat_or_title)
+        else:
+            raise Exception('Broken 2')
```
This needs fixed.

d0ugal commented on mkdocs/nav.py:
```diff
@@ -148,6 +186,63 @@ def __init__(self, title, url, path, url_context):
 
+        self.markdown, self.meta = self.load_markdown()
+
+    def __eq__(self, other):
```
It feels a bit like a hack adding this, but it made the tests much easier to do. Need to decide if we want to keep it or not.

d0ugal commented on mkdocs/serve.py:
```diff
@@ -71,7 +71,7 @@ def builder():
         )
         config['site_dir'] = tempdir
-        build(config, live_server=True)
+        build(config, live_server=True, clean_site_dir=True)
```
This is an incidental change and should be moved out.

d0ugal commented on mkdocs/tests/config/config_tests.py:
```diff
@@ -151,32 +149,33 @@ def test_default_pages_nested(self):
             })
             conf.validate()
+
             self.assertEqual([
```
This is so gross. Is there a better way to do this?

d0ugal commented on mkdocs/utils/mdutils.py:
```diff
@@ -0,0 +1,4 @@
+
+
+def get_markdown_title(markdown):
+    return
```
This needs to be implemented :smile:

d0ugal commented on mkdocs/nav.py:
```diff
@@ -26,6 +29,37 @@ def filename_to_title(filename):
+            pages_config[i] = _path_to_page(entry, use_directory_urls, docs_dir)
+            continue
+        elif not isinstance(entry, dict):
+            msg = ("Line in 'page' config is of type {0}, dict or string "
```
This is the error message we had before, it is terrible still. Should improve.

d0ugal commented on mkdocs/nav.py:
```diff
@@ -35,6 +69,9 @@ def __init__(self, pages_config, use_directory_urls=True):
         self.use_directory_urls = use_directory_urls
 
+        for page in self.pages:
+            page.url_context = self.url_context
```
Because the page object is created at a different time, we need to add this here. It seems quite hacky tho - maybe add a method to Page to set the context?

tkirill commented:
@d0ugal This is that required refactoring for #316?  If so then what is the status of this PR?  Is it waiting for review or there are some tasks to be done?  If there are any free tasks to be done to merge this PR I would be happy to do them.

waylan commented:
@tkirill, yes this is the refactoring for #316 among other things. I believe this is waiting on the issues mentioned in the comments above.

d0ugal commented:
@tkirill Yup, this re-factoring would enable #316 (it actually [almost includes the implementation](https://github.com/mkdocs/mkdocs/pull/627#discussion-diff-33427199)) and a number of other things we want to do. @waylan is correct that the above comments and check list in the description [here](https://github.com/mkdocs/mkdocs/pull/627#issue-87650999) are the outstanding tasks.

If you want to try any of them, I'd suggest making pull requests against my fork and then they can be merged in and included here. Stick to one per commit to make things easier to review if you can.

d0ugal commented:
Closing in favour of @tomchristie's work. https://github.com/mkdocs/mkdocs/tree/restructure
